### ISSUE TITLE
VmCluster update causes StatefulSet rolling-update timeout, removing all pods

### ISSUE THREAD

**Issue by antoinedeschenes** [VictoriaMetrics/operator#643]
If I make a modification to the VMCluster CRD, the underlying statefulset (in OnDelete mode) gets updated, and VMOperator deletes the first pod. However, it seems to timeout while waiting for the pod to turn ready, and starts deleting the following ones without waiting for the first one to become ready. It ends up scaling the vmstorage StatefulSet down to zero and making VM unavailable.

I'm still trying to figure out what's going on with the cluster, as pods don't get recreated by the stateful set afterwards, but I'm not sure operator should keep going with the rollout if deleted pods don't come back

```
{"level":"error","ts":1683827979.8579168,"logger":"client_utils","msg":"cannot get ready status for already updated pod","controller":"sts.rollingupdate","desiredVersion":"vmstorage-vmstack-754c8dcdb5","wasRecreated":false,"pod":"vmstorage-vmstack-1","error":"timed out waiting for the condition","stacktrace":"github.com/VictoriaMetrics/operator/controllers/factory/k8stools.performRollingUpdateOnSts\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/k8stools/sts.go:182\ngithub.com/VictoriaMetrics/operator/controllers/factory/k8stools.HandleSTSUpdate\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/k8stools/sts.go:75\ngithub.com/VictoriaMetrics/operator/controllers/factory.createOrUpdateVMStorage\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/vmcluster.go:286\ngithub.com/VictoriaMetrics/operator/controllers/factory.CreateOrUpdateVMCluster\n\tgithub.com/VictoriaMetrics/operator/controllers/factory/vmcluster.go:63\ngithub.com/VictoriaMetrics/operator/controllers.(*VMClusterReconciler).Reconcile\n\tgithub.com/VictoriaMetrics/operator/controllers/vmcluster_controller.go:78\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:121\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:320\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683827979.8740299,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"05ac6c8a-7105-41c5-9d29-9e0526baa5b3","error":"failed create or update vmcluster: cannot handle rolling-update on sts: vmstorage-vmstack, err: timed out waiting for the condition","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683827984.1136076,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"f27a0df5-7e49-468a-b36a-0c0b919f660d","error":"cannot update cluster status : Operation cannot be fulfilled on vmclusters.operator.victoriametrics.com \"vmstack\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683828085.5164416,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"fe2cd3e9-b2f2-4093-bdd5-ad0397039166","error":"cannot update cluster status : Operation cannot be fulfilled on vmclusters.operator.victoriametrics.com \"vmstack\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683828177.6376266,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"5ea6052f-fd76-482b-873b-b1b912fce708","error":"failed create or update vmcluster: cannot handle rolling-update on sts: vmstorage-vmstack, err: timed out waiting for the condition","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683828269.7419307,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"88865c4e-dc87-43a5-9fa0-58d2fba8e9c0","error":"failed create or update vmcluster: cannot handle rolling-update on sts: vmstorage-vmstack, err: timed out waiting for the condition","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683828601.3000138,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"2c8e3735-fd2c-4513-8dcb-9865377d7dc7","error":"cannot set expanding status for cluster: Operation cannot be fulfilled on vmclusters.operator.victoriametrics.com \"vmstack\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683831316.1564727,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"1c2ef829-e0fe-466f-951f-509a72bcc69c","error":"failed create or update vmcluster: cannot handle rolling-update on sts: vmstorage-vmstack, err: timed out waiting for the condition","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683831408.248902,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"48625b4c-755d-41cc-b015-2190c33b2ba1","error":"failed create or update vmcluster: cannot handle rolling-update on sts: vmstorage-vmstack, err: timed out waiting for the condition","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
{"level":"error","ts":1683831412.4389415,"msg":"Reconciler error","controller":"vmcluster","controllerGroup":"operator.victoriametrics.com","controllerKind":"VMCluster","VMCluster":{"name":"vmstack","namespace":"victoria-metrics"},"namespace":"victoria-metrics","name":"vmstack","reconcileID":"d3e802bb-325f-4b77-bee1-2bf1a47358bc","error":"cannot set expanding status for cluster: Operation cannot be fulfilled on vmclusters.operator.victoriametrics.com \"vmstack\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234"}
```

**Comment by f41gh7** [comment]
Hello, is it latest operator version? The same issue was fixed a couple releases ago.

**Comment by antoinedeschenes** [comment]
Happened again on another GKE cluster with 0.33.0. 

Should the operator stop deleting sts pods if the pod count isn't right?

I believe we might be ending up in the `if len(podList.Items) != neededPodCount` case, but it's not preventing a further pod deletion.

https://github.com/VictoriaMetrics/operator/blob/7c146e5cea3de30843448a3132c7d8b9a41a737a/controllers/factory/k8stools/sts.go#L123C1-L126

**Comment by f41gh7** [comment]
It looks like a source of an issue. I've added special case for it at the related commit.

 Operator may delete pods, if it's definition incorrectly configured and kubernetes controller-manager cannot create it.
 In this case, operator will fail fast and prevent possible pod deletion.

**Comment by antoinedeschenes** [comment]
Awesome, thanks!

**Comment by f41gh7** [comment]
An issue must be fixed at [v0.34.1 release](https://github.com/VictoriaMetrics/operator/releases/tag/v0.34.1)
